configuration_go/k8sutil/genericdeployment.go

```diff
@@ -0,0 +1,154 @@
+package k8sutil
+
+import (
+	monv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// ExtraConfig represents the configuration required to add extra containers, volumes
+// service and servicemonitor ports to a particular Deployment/StatefulSet.
+type ExtraConfig struct {
+	Sidecars                      []corev1.Container
+	AdditionalPodVolumes          []corev1.Volume
+	AdditionalServicePorts        []corev1.ServicePort
+	AdditionalServiceMonitorPorts []monv1.Endpoint
+}
+
+// DeploymentGenericConfig represents certain config fields
+// that might be useful to add/override in a Deployment. It contains
+// fields of both DeploymentSpec and PodSpec.
+// It also has method defined for overriding any default values.
+type DeploymentGenericConfig struct {
```

Comment:
I feel we could do something simpler here by avoiding filling transitory fields that will then be mapped to the `appsv1.Deployment` struct. Couldn't we directly manipulate this target structure here and avoid having to map everything back to it in the `Manifests()` methods used in `abstr/kubernetes`.
Also, we are coupling this implementation with the `Manifests()` methods as additional capability provided here will probably need modifications in `Manifests()` to map the fields in the target `appsv1.Deployment` struct.
I might have overlooked some details preventing this approach. Feel free to discard this proposition.

Reply:
Thanks! This seems doable and a good idea, will iterate on this in a later PR 🙂